colcon_core/package_augmentation/python.py

```diff
@@ -80,6 +80,15 @@ def extract_dependencies(options):
         for dep in options.get(option_name, []):
             dependencies[dependency_type].add(
                 create_dependency_descriptor(dep))
+
+    extras_mapping = {
+        'test': 'test',
+    }
+    extras_require = options.get('extras_require', {})
+    for option_name, dependency_type in extras_mapping.items():
```

Comment:
Alright, I gave it a shot. It makes it a little harder to read, though. Let me know what you think.

Reply:
I think is really good to avoid the code duplication in this case, so +1. It is true that is hard to read. Not sure if much better but maybe applying:

```diff
diff --git a/colcon_core/package_augmentation/python.py b/colcon_core/package_augmentation/python.py
index 5bd1e9c..a0b9701 100644
--- a/colcon_core/package_augmentation/python.py
+++ b/colcon_core/package_augmentation/python.py
@@ -75,19 +75,16 @@ def extract_dependencies(options):
         'tests_require': 'test',
     }
     dependencies = {}
-    _map_dependencies(mapping, options, dependencies)
-
-    extras_mapping = {
-        'test': 'test',
-    }
+    _map_dependencies(options, mapping, dependencies)
+    # include options.extras_require.test in test dependencies
     _map_dependencies(
-        extras_mapping, options.get('extras_require', {}),
+        options.get('extras_require', {}), {'test': 'test'},
         dependencies)
 
     return dependencies
 
 
-def _map_dependencies(mapping, options, dependencies):
+def _map_dependencies(options, mapping, dependencies):
     for option_name, dependency_type in mapping.items():
         dependencies.setdefault(dependency_type, set())
         for dep in options.get(option_name, []):
```
will make it to be:

```python
    mapping = {
        'setup_requires': 'build',
        'install_requires': 'run',
        'tests_require': 'test',
    }
    dependencies = {}
    _map_dependencies(options, mapping, dependencies)
    # include options.extras_require.test in test dependencies
    _map_dependencies(
        options.get('extras_require', {}), {'test': 'test'},
        dependencies)
```
I leave it to your criteria, I think is good enough.

Reply:
I took some of this suggestion, but I also extended the mapping to cover a couple of other common `extras_require` names that indicate test dependencies.